logstash-core/src/main/java/org/logstash/instrument/metrics/PipelineMetrics.java

```diff
@@ -0,0 +1,64 @@
+package org.logstash.instrument.metrics;
+
+import org.logstash.Timestamp;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Created by andrewvc on 5/30/17.
+ */
+public class PipelineMetrics {
+    private final String pipelineId;
+    private final MetricFactory metricFactory;
+    private final List<String> namespaceRoot;
+    private final List<String> reloadsNamespace;
+    private final Counter reloadSuccesses;
+    private final Counter reloadFailures;
+    private final Gauge<Map> reloadLastError;
+    private final Gauge<Timestamp> lastSuccessTimestamp;
+    private final Gauge<Timestamp> lastFailureTimestamp;
+    private final AgentMetrics agentMetrics;
+
+    public PipelineMetrics(String pipelineId, MetricFactory metricFactory, AgentMetrics agentMetrics) {
+        this.pipelineId = pipelineId;
+        this.metricFactory = metricFactory;
+        this.agentMetrics = agentMetrics;
+        this.namespaceRoot = Arrays.asList("stats", "pipelines", pipelineId);
+
+        this.reloadsNamespace = Stream.concat(namespaceRoot.stream(), Stream.of("reloads")).collect(Collectors.toList());
+        this.reloadSuccesses = metricFactory.makeLongCounter(reloadsNamespace, "successes", 0);
+        this.reloadFailures = metricFactory.makeLongCounter(reloadsNamespace, "failures", 0);
+
+        this.lastSuccessTimestamp = metricFactory.makeGauge(reloadsNamespace, "last_success_timestamp", null);
+        this.lastFailureTimestamp = metricFactory.makeGauge(reloadsNamespace, "last_failure_timestamp", null);
+        this.reloadLastError = metricFactory.makeGauge(reloadsNamespace, "last_error", null);
+    }
+
+    public void reloadSuccess() {
+        this.agentMetrics.reloadSuccess();
```

Comment:
I think the observer pattern would be cleaner here, maybe I should refactor so we have an `AgentObserver` and a `PipelineObserver` listening to some intermediate object? Maybe too much for an initial refactor, I dunno.